middle_end/flambda/types/inlining/inlining_arguments.ml

```diff
@@ -14,65 +14,158 @@
+    indirect_cost = cost_f !Clflags.inline_indirect_cost ~round;
+    small_function_size = cost_i !Clflags.inline_small_function_size ~round;
+    large_function_size = cost_i !Clflags.inline_large_function_size ~round;
+    threshold = cost_f !Clflags.inline_threshold ~round;
+  }
+end
+
+type t = Args.t Or_unknown.t
+
+let unknown = Or_unknown.Unknown
+
+let print ppf = Or_unknown.print Args.print ppf
 
 let max_inlining_depth = function
-  | Unknown -> 0
-  | O1 -> 1
-  | O2 -> 2
-  | O3 -> 3
-  | Custom t -> t.max_inlining_depth
-
-let equal t1 t2 =
-  max_inlining_depth t1 = max_inlining_depth t2
-
-let compress t =
-  match t with
-  | O1 | O2 | O3 | Unknown -> t
-  | t ->
-    if equal t O1 then O1
-    else if equal t O2 then O2
-    else if equal t O3 then O3
-    else t
-
-let meet_custom t1 t2 =
-  let max_inlining_depth =
-    if (max_inlining_depth t1) <= (max_inlining_depth t2) then
-      max_inlining_depth t1
-    else
-      max_inlining_depth t2
-  in
-  Custom { max_inlining_depth }
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ max_inlining_depth; _ } ->
+    max_inlining_depth
 
-let meet t1 t2 =
+let call_cost = function
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ call_cost; _ } ->
+    call_cost
+
+let alloc_cost = function
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ alloc_cost; _ } ->
+    alloc_cost
+
+let prim_cost = function
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ prim_cost; _ } ->
+    prim_cost
+
+let branch_cost = function
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ branch_cost; _ } ->
+    branch_cost
+
+let indirect_cost = function
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ indirect_cost; _ } ->
+    indirect_cost
+
+let small_function_size = function
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ small_function_size; _ } ->
+    small_function_size
+
+let large_function_size = function
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ large_function_size; _ } ->
+    large_function_size
+
+let threshold = function
+  | Or_unknown.Unknown -> assert false
+  | Or_unknown.Known Args.{ threshold; _ } ->
+    threshold
+
+let meet (t1 : _ Or_unknown.t) (t2 : _ Or_unknown.t) : t =
   match t1, t2 with
   | Unknown, Unknown -> Unknown
-  | Unknown, (O1 | O2 | O3 | Custom _) -> t2
-  | (O1 | O2 | O3 | Custom _), Unknown -> t1
-
-  | O1, O1 | O1, (O2 | O3) | (O2 | O3), O1 -> O1
-  | O2, O2 | O3, O2 | O2, O3 -> O2
-  | O3, O3 -> O3
-  | Custom _, _ | _, Custom _ -> meet_custom t1 t2
-
-let create ~round:_ =
-  compress (Custom {
-    max_inlining_depth = !Clflags.Flambda.Expert.max_inlining_depth
-  })
+  | Known _, Unknown -> t1
+  | Unknown, Known _ -> t2
+  | Known args1, Known args2 ->
+    if Args.(<=) args1 args2 then
+      t1
+    else if Args.(<=) args2 args1 then
+      t2
+    else
+      Known (Args.meet args1 args2)
```

Comment:
These conditionals seem wrong maybe?  Could this just call `meet`, and that's all?

Reply:
I've added a comment to explain a bit more. Basically if we know that one set of arguments will inline strictly less than the other then meet will just return it. In that case we can skip calling [Args.meet] and allocating a new [Args.t].